### ISSUE TITLE
Wrong coefficients coupling the eyes' vergence?

### ISSUE THREAD

**Issue by pattacini** [robotology/robots-configuration#228]
Given that the vergence `Vg` amounts to the difference between the single eyes' joints `L-R` (see the [wiki][1]), I think that the [values contained in the eyes' coupling matrices][2] should be **`-1.0 1.0`** instead of `-0.5  0.5`.

Differently, as `Vs = (L + R) / 2` it turns out that the line above reporting `0.5  0.5` is spot on.

Can this reasoning be assumed correct?

This is somehow linked to https://github.com/robotology/icub-firmware/issues/71.

cc @davidetome @randaz81 @ale-git @traversaro @plinioMoreno 

[1]: https://wiki.icub.org/wiki/Vergence,_Version_and_Disparity
[2]: https://github.com/robotology/robots-configuration/blob/master/iCubGenova04/hardware/mechanicals/head-eb21-j2_5-mec.xml#L45

**Comment by pattacini** [comment]
Any comments on this @ale-git @randaz81 ?

**Comment by pattacini** [comment]
Discussed this with @randaz81 today, who agrees with what said above.
However, the wrong matrix coefficients get subsumed by the PID gains, hence there ain't actually any real control problem.

Given that it's more correct to have `-1.0 1.0`, it'd be worth updating the conf files anyway.

**Comment by pattacini** [comment]
Apparently, also matrixE2J should contain coupling terms!

**Comment by pattacini** [comment]
/remind September 03 let's address this once for good.

**Comment by pattacini** [comment]
Link in the OP fixed.

**Comment by pattacini** [comment]
To sum up, here's the list of things that we need to do:
- Stick to one iCub with eyes; most likely iCubGenova11.
- Update iCubGenova11 following these [instructions](https://github.com/robotology/robots-configuration/issues/228#issue-783526278).
- Test that the eyes will still move correctly w/o big overshoot.
- Update all the remaining maintained robots.

**Comment by martinaxgloria** [comment]
Hi @pattacini,

as you suggested, I booked iCubGenova11 to do the tests with the new values of eyes' vergence. I took a video of the eyes' movement after the upload:


https://github.com/robotology/robots-configuration/assets/114698424/1adeb6c5-f5f3-4a80-90ab-dce5e693c6f5

I have little experience with it, but the movement seems smooth, also compared to the previous version with the _wrong_ vergence values:


https://github.com/robotology/robots-configuration/assets/114698424/016518b5-3a1f-4f85-969a-ff8347b84727

Do you agree? If yes, I'll proceed with the update of all the maintained robots.

**Comment by martinaxgloria** [comment]
Just to clarify, I only update [this line](https://github.com/robotology/robots-configuration/blob/master/robots-icebox/iCubGenova04/hardware/mechanicals/head-eb21-j2_5-mec.xml#L45) with `-1.0 1.0` instead of `-0.5 0.5`.

**Comment by pattacini** [comment]
Great, @martinaxgloria 👍🏻 

That's all correct!

The videos are nice. However, given the control context, I would expand on the validation phase by including some data acquisitions of the commands and joint feedback. These traces would help us appreciate small details that are not visible in the videos, like overshoots, rising times...

**Comment by martinaxgloria** [comment]
You're right @pattacini! I'll do the acquisitions and I'll report here the results.

**Comment by pattacini** [comment]
You can talk to @davidetome and/or @mfussi66, who do have some code snippets for logging both commands and feedback you may consider reusing.

**Comment by martinaxgloria** [comment]
Today I plotted the data I've acquired on iCubGenova11 after the change in the vergence coupling matrix and here they are:

![image](https://github.com/robotology/robots-configuration/assets/114698424/a93368be-34a3-46ce-ba24-8235769a3cc5)

I hope that these quantities are the ones that @pattacini referred to. In particular, I plotted the `encoder`, `motor encoder`, `pwm` and `current` values for the `eyes_tilt`, `eyes_vers` and `eyes_verg` joints. These values were obtained by setting a sequence from the yarpmotorgui from the home position, to the upper limit, then the lower limit and then home again for the three eyes' joints. I set a time of 5 sec for each movement (e.g. from home to upper limit in 5 sec). Then I start the acquisition and I ran the code to save the values in order to plot them at the end.

The only particular thing that stands out is the peak of current and duty cycle percentage around the second 45 that corresponds to the `eyes_verg` joint reaching its upper limit. In your opinion, is it consistent with the movement?

cc @pattacini @mfussi66

**Comment by pattacini** [comment]
Great that you managed to plot these results 👍🏻 
A couple of important points though:
- The joint references are missing.
- The 5-second movements are far slower than the dynamics we're interested in with the eyes. Eyes need to move much much faster, meaning fractions of a second. For example, let's stick to 0.3 s.

**Comment by martinaxgloria** [comment]
Yesterday I repeated the tests acquiring the joint references and the other quantities with much faster movements (0.3 sec per movement). To make a comparison, I also do the tests and the acquisitions before the change:

![old](https://github.com/robotology/robots-configuration/assets/114698424/0b66bbe4-68af-4b80-988d-5236f42ab94d)
and the comparison between the joint encoder angles and the reference values:

![encoder_pid_old](https://github.com/robotology/robots-configuration/assets/114698424/a4390bd2-a926-426b-bd71-6355768ac68c)

After the changes, instead:

![new](https://github.com/robotology/robots-configuration/assets/114698424/2e00e33f-9805-43e5-b355-24f970c65432)

![encoder_pid_new](https://github.com/robotology/robots-configuration/assets/114698424/82fc6d5a-3590-4387-85a3-e151b832def8)

@mfussi66 noticed that the sampling time it's high (when I did again the acquisitions maybe I accidentally change this parameter and I didn't notice it). So far, I decided to post here the results, but I'll do again the acquisitions and the plots with more samplings to have a more accurate analysis.

cc @pattacini

**Comment by pattacini** [comment]
Overall, I'm happy with the results and for me it's green light 🟢
Thus, we can proceed with the update of `robots-configuration`.

>@mfussi66 noticed that the sampling time it's high (when I did again the acquisitions maybe I accidentally change this parameter and I didn't notice it). So far, I decided to post here the results, but I'll do again the acquisitions and the plots with more samplings to have a more accurate analysis.

It seems so, yeah. However, what was the sample time you used?

**Comment by martinaxgloria** [comment]
> It seems so, yeah. However, what was the sample time you used?

I computed it from the acquired samples and it is 0.1s

**Comment by pattacini** [comment]
> > It seems so, yeah. However, what was the sample time you used?
> 
> I computed it from the acquired samples and it is 0.1s

Ok, it's fairly low, but the graphs above demonstrate that the reference is well-tracked anyhow.

For the record, good a sample time when dealing with control issues is 10 ms at the YARP layer.

**Comment by martinaxgloria** [comment]
Yes, thanks for the clarification @pattacini!
So, can I proceed with the update of robots-configuration or it's better to re-do the acquisitions with the right sample time before?

**Comment by pattacini** [comment]
>So, can I proceed with the update of robots-configuration or it's better to re-do the acquisitions with the right sample time before?

Let's proceed with robots-configuration 👍🏻

**Comment by martinaxgloria** [comment]
PR here:

- https://github.com/robotology/robots-configuration/pull/574
